### PR TITLE
[AIRFLOW-5902] avoid unnecessary sleep to maintain local task job heart rate

### PR DESCRIPTION
**Pull request by houqp**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-XXX
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

After a heartbeat, time since last hearbeat will always be lower than heart rate, so the subsequent sleep is unnecessary. It also spams the task log with lots of info messages.

Sleep to maintain heart rate is already done by the `heartbeat()` call.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

File: airflow/jobs/base_job.py [hunk 1]
```diff
@@ -171,17 +171,14 @@ def heartbeat(self):
             if self.state == State.SHUTDOWN:
                 self.kill()
 
-            is_unit_test = conf.getboolean('core', 'unit_test_mode')
```
Comment (KevinYang21) [inline airflow/jobs/base_job.py:174]: Do you mind also include the reason why do we want to remove the `is_unit_test` check  please?

Comment (houqp) [reply airflow/jobs/base_job.py:174]: This if condition was added to speed up the tests, however, there is a much better and more maintainable way to achieve the same effect, which is mock the sleep call from unit tests. By doing that, we get the following benefits:

1. in unit test, we can now check and assert whether sleep has been called with the right argument
2. avoid unnecessary dictionary lookup and if statement check at runtime
3. avoid the anti-pattern where unit test logic is mingled with business logic. unit test should be as transparent to the actual code as possible

I noticed we are adopting this pattern in 2 other places as well and I really think we should replace all of them with proper mocking in the test and keep the business code clean.

Comment (potiuk) [reply airflow/jobs/base_job.py:174]: I agree in principle @houqp. But should not we remove the check and add mocking it in the same commit? I vaguely remember that this single "if" caused few minutes shorter CI build - so if we remove it now, everyone will be impacted.

Comment (houqp) [reply airflow/jobs/base_job.py:174]: @potiuk I have already added the mock in this PR. Or are you thinking of separating out the mock change into its own commit?

Comment (houqp) [reply airflow/jobs/base_job.py:174]: I am thinking of patching sleep globally for all unit tests, so we won't have to worry about it going forward. is this the kind of change you all comfortable with?

Comment (houqp) [reply airflow/jobs/base_job.py:174]: @KevinYang21 @potiuk is there anything you want me to change?

Comment (ashb) [reply airflow/jobs/base_job.py:174]: > I am thinking of patching sleep globally for all unit tests, so we won't have to worry about it going forward. is this the kind of change you all comfortable with?

No, sometimes we _need_ sleep in our tests to give a sub-process time to do start up etc.(mostly tests, not in the airflow core), or in the dags to "do nothing for a long time" so this would break all sorts of things.

**Comment by jj-ookla** [comment]
If the unit test changes are not approved, I would really like to see the removal of heart beat logging make its way into the next release. This has considerably inflated our log files.

https://github.com/apache/airflow/blob/master/airflow/jobs/local_task_job.py#L118

File: airflow/jobs/base_job.py (hunk 1 above)
Comment (houqp) [reply airflow/jobs/base_job.py:174]: That's fair, I will leave it to test level like it is right now then.

**Comment by houqp** [comment]
@jj-ian this exactly how I came to this patch :) Feel free to apply this patch to your cluster, we have been running airflow with it in production for a week now.

File: airflow/jobs/base_job.py (hunk 1 above)
Comment (houqp) [reply airflow/jobs/base_job.py:174]: @KevinYang21 @potiuk @ashb are you cool with the unit test changes I made in this PR? If not, please feel free to comment. This is a pretty annoying bug and renders all task logs not readable.

**Review by KevinYang21 — APPROVED** [review]

File: airflow/models/taskinstance.py [hunk 2]
```diff
@@ -1320,9 +1320,9 @@ def set_duration(self):
 
     def xcom_push(
             self,
-            key,
-            value,
-            execution_date=None):
+            key: str,
+            value: Any,
+            execution_date: Optional[datetime] = None) -> None:
```
Comment (ashb) [inline airflow/models/taskinstance.py]: Are these changes required to make this PR pass tests? If not it is preferred to not change files that you don't have to for the subject of the PR.

Comment (houqp) [reply airflow/models/taskinstance.py]: I have removed the type annotation.

**Comment by houqp** [comment]
@ashb type annotation additions are not moved to https://github.com/apache/airflow/pull/6674.

File: airflow/jobs/base_job.py (hunk 1 above)
Comment (ashb) [inline airflow/jobs/base_job.py:174]: Does anything else actually use this setting anymore, or could we remove it now?

File: airflow/jobs/local_task_job.py [hunk 3]
```diff
@@ -112,13 +111,6 @@ def signal_handler(signum, frame):
                                            "exceeded limit ({}s)."
                                            .format(time_since_last_heartbeat,
                                                    heartbeat_time_limit))
-
-                if time_since_last_heartbeat < self.heartrate:
```
Comment (ashb) [inline airflow/jobs/local_task_job.py:116]: Ah - the the check above is done _after_ calling `self.heartbeat()` which means it may have just updated the `latest_heartbeat` -- should that check be done before calling `heartbeat()`?

File: tests/jobs/test_local_task_job.py [hunk 4]
```diff
@@ -43,6 +43,9 @@
 class TestLocalTaskJob(unittest.TestCase):
     def setUp(self):
         clear_db_runs()
+        patcher = patch('airflow.jobs.base_job.sleep')
```
Comment (ashb) [inline tests/jobs/test_local_task_job.py:46]: What about scheduler_job and backill_job -- shouldn't we need to patch those too?

Comment (ashb) [reply tests/jobs/test_local_task_job.py:46]: I guess not? - it doesn't look like the tests take any longer without it (it's within the normal variance of tests I think)

File: airflow/jobs/base_job.py (hunk 1 above)
Comment (houqp) [reply airflow/jobs/base_job.py:174]: This is still used in other tests unfortunately. But this PR should setup the example on how to remove or avoid using it in other places going forward.

File: airflow/jobs/local_task_job.py (hunk 3 above)
Comment (houqp) [reply airflow/jobs/local_task_job.py:116]: heartbeat already has this check to maintain the heart rate, so this logic is redundant for all `heartbeat()` callers. It's safe to just remove it :) The test covers this to make sure heart rate is still maintained without this check.

File: tests/jobs/test_local_task_job.py (hunk 4 above)
Comment (houqp) [reply tests/jobs/test_local_task_job.py:46]: yeah, this test doesn't cover the code path that does the actual sleep call in those two classes. we should definitely patch them if a test runs through those code path going forward.
